internal/handlers/common/filter.go

```diff
@@ -55,85 +55,191 @@ func FilterDocument(doc, filter *types.Document) (bool, error) {
 		}
 
-		// we pass the path without the last key because we want {fieldN: *someValue*}, not just *someValue*
-		docValue, err := doc.GetByPath(path.TrimSuffix())
-		if err != nil {
-			return false, nil // no error - the field is just not present
+		if filterSuffix, docs = getDocumentsAtSuffix(doc, path); len(docs) == 0 {
+			// When no document is found at suffix, use an empty one.
+			// So operators such as $nin is applied to the empty document.
+			docs = append(docs, types.MakeDocument(0))
 		}
+	}
 
-		switch docValue := docValue.(type) {
+	if strings.HasPrefix(filterKey, "$") {
+		// {$operator: filterValue}
+		return filterOperator(doc, filterKey, filterValue)
+	}
+
+	for _, doc := range docs {
+		switch filterValue := filterValue.(type) {
 		case *types.Document:
-			doc = docValue
-			filterKey = path.Suffix()
+			// {field: {expr}} or {field: {document}}
+			ok, err := filterFieldExpr(doc, filterKey, filterSuffix, filterValue)
+			if err != nil {
+				return false, err
+			}
+
+			if ok {
+				return true, nil
+			}
+
+			// doc did not match filter, continue next iteration.
 		case *types.Array:
-			index, err := strconv.Atoi(path.Suffix())
+			// {field: [array]}
+			docValue, err := doc.Get(filterSuffix)
 			if err != nil {
-				return false, nil
+				continue // no error - the field is just not present
 			}
 
-			if docValue.Len() == 0 || docValue.Len() <= index {
-				return false, nil
+			if result := types.Compare(docValue, filterValue); result == types.Equal {
+				return true, nil
+			}
+
+			// doc did not match filter, continue next iteration.
+		case types.Regex:
+			// {field: /regex/}
+			docValue, err := doc.Get(filterSuffix)
+			if err != nil {
+				continue // no error - the field is just not present
 			}
 
-			value, err := docValue.Get(index)
+			ok, err := filterFieldRegex(docValue, filterValue)
 			if err != nil {
 				return false, err
 			}
 
-			doc = must.NotFail(types.NewDocument(filterKey, value))
+			if ok {
+				return true, nil
+			}
+
+			// doc did not match filter, continue next iteration.
+		default:
+			// {field: value}
+			docValue, err := doc.Get(filterSuffix)
+			if err != nil {
+				// comparing not existent field with null should return true
+				if _, ok := filterValue.(types.NullType); ok {
+					return true, nil
+				}
+
+				continue // no error - the field is just not present
+			}
+
+			if result := types.Compare(docValue, filterValue); result == types.Equal {
+				return true, nil
+			}
 		}
 	}
 
-	if strings.HasPrefix(filterKey, "$") {
-		// {$operator: filterValue}
-		return filterOperator(doc, filterKey, filterValue)
-	}
+	// If we got here, it means that none of the documents matched the filter.
+	return false, nil
+}
 
-	switch filterValue := filterValue.(type) {
-	case *types.Document:
-		// {field: {expr}} or {field: {document}}
-		return filterFieldExpr(doc, filterKey, filterValue)
+// getDocumentsAtSuffix finds documents at the suffix of the given path.
```

Comment:
Let's add a comment that we go iteratively through the document to find all those matches.

Reply:
I added comment and sample example cases in doc string.